python_onnxifi/data_conversion.cc

```diff
@@ -0,0 +1,298 @@
+      throw std::runtime_error("Dispatch received non-numeric data type");   \
+    }                                                                        \
+  }
+
+template <typename onnx_type, typename unused>
+void DataConversion::getBufferSize(uint64_t& buffer_size,
+                                   const uint64_t* shape,
+                                   uint32_t dimensions) {
+  auto numElements = std::accumulate(shape, shape + dimensions, (uint64_t)1,
+                                     std::multiplies<uint64_t>());
+  buffer_size = sizeof(onnx_type) * numElements;
+}
+
+void DataConversion::updateOutputDescriptorsData(ModelProto& model) {
+  output_descriptors_data_.clear();
+  ::ONNX_NAMESPACE::shape_inference::InferShapes(model);
+  const auto& outputs = model.graph().output();
+  for (const auto& vi : outputs) {
+    DescriptorData outputData;
+    outputData.name = vi.name();
+    outputData.descriptor.name = outputData.name.c_str();
+    if (!vi.type().tensor_type().has_elem_type()) {
+      throw std::runtime_error(
+          "Non-static ModelProto: Output data type not found");
+    }
+    outputData.descriptor.dataType = vi.type().tensor_type().elem_type();
+    outputData.descriptor.dimensions =
+        vi.type().tensor_type().shape().dim_size();
+    ;
+    for (auto i = 0; i < outputData.descriptor.dimensions; ++i) {
+      const auto& dim = vi.type().tensor_type().shape().dim(i);
+      if (!dim.has_dim_value()) {
+        throw std::runtime_error(
+            "Non-static ModelProto: Output shape dimension not found");
+      }
+      outputData.shape.push_back(dim.dim_value());
+    }
+    outputData.descriptor.shape = outputData.shape.data();
+    outputData.descriptor.memoryType =
+        ONNXIFI_MEMORY_TYPE_CPU;  // TODO: Expand memory types?
+    uint64_t buffer_size;
+    DISPATCH_OVER_NUMERIC_DATA_TYPE(
+        outputData.descriptor.dataType, getBufferSize, buffer_size,
+        outputData.descriptor.shape, outputData.descriptor.dimensions)
+    outputData.buffer.resize(buffer_size);
+    outputData.descriptor.buffer =
+        reinterpret_cast<onnxPointer>(outputData.buffer.data());
+    output_descriptors_data_.push_back(std::move(outputData));
+  }
+}
+
+template <typename onnx_type, typename py_type>
+void DataConversion::addNumpyArray(py::dict& numpyArrays,
+                                   const DescriptorData& dd) {
+  auto numElements = std::accumulate(dd.shape.begin(), dd.shape.end(),
+                                     (uint64_t)1, std::multiplies<uint64_t>());
+  py_type intermediateBuffer[numElements];
+  onnx_type* oldBuffer = reinterpret_cast<onnx_type*>(dd.descriptor.buffer);
+  std::copy(oldBuffer, oldBuffer + numElements, intermediateBuffer);
+  std::vector<ptrdiff_t> shape(dd.descriptor.shape,
+                               dd.descriptor.shape + dd.descriptor.dimensions);
+  auto numpyArray = py::array_t<py_type>(shape, intermediateBuffer);
```

Comment:
I'm not very familiar with numpy API. Here do we make a copy of `intermediateBuffer` or we just grab its pointer?

Reply:
I think initially they always made a copy (https://github.com/pybind/pybind11/issues/27). But looking at the code (https://github.com/pybind/pybind11/blob/master/include/pybind11/numpy.h#L553) it seems like the default is to not copy now?